File: Coding Interview Workshops (CIW)/Week 4 - Hash Tables/Code/InterviewQs/TwoSum.py
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        """
        for i, num1 in enumerate(nums):
            for j, num2 in enumerate(nums[i+1::]):
                if ((num1 + num2) == target):
                    return [i, j+i+1]
        """

        hashT = {}

        for i, num1 in enumerate(nums):
            comp = target - num1
            if (comp in hashT):
                return [hashT[comp], i]
            else:
                hashT[num1] = i

            

File: Coding Interview Workshops (CIW)/Week 04 - Hash Tables/Code/InterviewQs/TwoSum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */

// APC
// Assumptions:
// Each input has one solution
// Can't use the same element twice
// Return answer in any order
// Efficient time wise solution
// nums is small
// int and target are big

// Plan:
// 1. Obvious solution: Look at each number, loop through rest try and form target [O(n2)]
// 2. Better solution: Searching more efficient with Hash Tables
// for each element: {key = value, index}, then check if the complement is in the map already
// is [2] in the list -> yes / no
// O(n)


var twoSum = function (nums, target) {
    let valueMap = new Map();

    for (let i = 0; i < nums.length; i++) {
        let comp = target - nums[i];
        if (valueMap.has(comp)) {
            return [i, valueMap.get(comp)];
        }
        valueMap.set(nums[i], i)
    }
};